### PR TITLE
#156 implement rest env api in ruby

### PR DESCRIPTION
**Pull request by cdeler**
@lopopolo I had come back from vacation and fixed your remarks (from [this](https://github.com/artichoke/artichoke/issues/156#issuecomment-521088939) comment).

### CODE REVIEW

File: artichoke-backend/src/extn/core/env/env.rb [hunk 1]
```diff
@@ -1,3 +1,119 @@
...
+  end
+
+  def to_hash
+    to_h
+  end
+
+  def shift
+    envs = to_h
+
+    return nil if envs.nil?
```
Comment (lopopolo) [inline artichoke-backend/src/extn/core/env/env.rb]: don't we also want to return nil if `envs.empty?`

File: artichoke-backend/src/extn/core/env/env.rb [hunk 2]
```diff
@@ -1,3 +1,119 @@
 # frozen_string_literal: true
 
-ENV = EnvClass.new
+class EnvClass
+  def assoc(name)
+    value = self[name]
+
+    value.nil? ? nil : [name, self[name]]
```
Comment (lopopolo) [inline artichoke-backend/src/extn/core/env/env.rb]: can we structure this as:

```ruby
return nil if value.nil?
[name, value]
```

File: artichoke-backend/src/extn/core/env/env.rb [hunk 3]
```diff
@@ -1,3 +1,119 @@
...
+    self[name] = nil
+    yield name if block_given?
+
+    value
+  end
+
+  def delete_if
+    return to_enum(:delete_if) unless block_given?
+
+    each do |key, value|
```
Comment (lopopolo) [inline artichoke-backend/src/extn/core/env/env.rb]: Are we missing a definition of `each`?

`clear` calls `to_h.each`. Should we be doing that here?

File: artichoke-backend/src/extn/core/env/env.rb (hunk 1 above)
Comment (cdeler) [reply artichoke-backend/src/extn/core/env/env.rb]: > Removes an environment variable name-value pair from ENV and returns it as an Array. Returns nil if when the environment is empty.
(c) https://ruby-doc.org/core-2.6/ENV.html#method-c-shift

Comment (lopopolo) [reply artichoke-backend/src/extn/core/env/env.rb]: Does `to_h` on an empty `ENV` return nil or an empty hash? An empty hash is not `nil`.

File: artichoke-backend/src/extn/core/env/env.rb (hunk 3 above)
Comment (cdeler) [reply artichoke-backend/src/extn/core/env/env.rb]: I'm sorry, I'm not familiar with ruby (and didn't check it), it's my fault.

File: artichoke-backend/src/extn/core/env/env.rb (hunk 1 above)
Comment (lopopolo) [reply artichoke-backend/src/extn/core/env/env.rb]: oops didn't see the thumbs up 😅

**Comment by lopopolo** [comment]
I've updated the checklist with the current state https://github.com/artichoke/artichoke/issues/156#issuecomment-521088939

File: artichoke-backend/src/extn/core/env/env.rb (hunk 1 above)
Comment (cdeler) [reply artichoke-backend/src/extn/core/env/env.rb]: I'm sorry, I did agree with you, I've rewritten this line

File: artichoke-backend/src/extn/core/env/env.rb [hunk 4]
```diff
@@ -1,3 +1,121 @@
 # frozen_string_literal: true
 
-ENV = EnvClass.new
+class EnvClass
+  def assoc(name)
+    value = self[name]
+
+    return nil if value.nil?
+
+    [name, self[name]]
```
Comment (lopopolo) [inline artichoke-backend/src/extn/core/env/env.rb]: @cdeler this comment might have gotten lost in the last review. Can we replace the call to `self[name]` here with the value we already pulled out of the env?

I'm worried about a race as we call back into Rust as Artichoke gains support for multi-threading.

Comment (cdeler) [reply artichoke-backend/src/extn/core/env/env.rb]: Done

**Pull request merged**